python/ray/data/_internal/block_list.py

```diff
@@ -15,19 +15,26 @@ class BlockList:
     change after execution due to block splitting.
     """
 
-    def __init__(self, blocks: List[ObjectRef[Block]], metadata: List[BlockMetadata]):
+    def __init__(
+        self,
+        blocks: List[ObjectRef[Block]],
+        metadata: List[BlockMetadata],
+        created_by_pipeline: bool = False,
```

Comment:
Is there a better name for this? e.g., `ephemeral` or `can_delete_once_consumed`?

Reply:
Or `lazily_created` or `pinned`? This will be consistent with the end state where if we know it's lazily created (either by Dataset or DatasetPipeline), it should be safe to clear, since lazy ds/dsp will recompute from the beginning (external files or pinned memory blocklist) if get accessed again ("action" APIs get called again). In other words, the logical owner of those blocklists is each single invocation of "action" APIs.

Reply:
Changed to "consumable", so make a distinction of consumable v.s. non-consumable blocklists. The consumable blocklists are like current DatasetPipeline, which can consume the data (i.e. disruptive read). LMK if that makes sense, or any preferred name/concept for this.

Reply:
Why not "movable", since that's what we're doing here, right?

Reply:
We do produce new blocks in stage/transform so I think move is not quite accurate.

Reply:
Alternative: owned_by_dataset v.s. owned_by_consumer (the consumer can be an action API at Dataset or DatasetPipeline).
This looks at all blocklists from a perspective of logical owners. The consumer-owned memory means they are created for each individual consumer, and for the same consumer we can eagerly release memories (they won't be read by other consumers). This will also future-proof as we further revamp the execution model.

Reply:
I quite like `owned_by_consumer`! It seems much more self-explanatory than consumable, which is a little vague.